### PR TITLE
Fix session report bug when database disabled.

### LINKED ISSUE #10129: Metasploit no longer reports when it receives a session
I noticed starting about a week ago that Metasploit no longer reports when it receives a session.

At first this wasn't an issue, as I was testing a staged payload, so `Sending stage (x bytes)` in console was a dead give away.

However, it's a problem. Observe:

```
msf5 exploit(multi/handler) > run

[*] Started reverse TCP handler on 10.1.1.197:1337 

id
uid=501(qq) gid=20(staff) groups=20(staff),401(com.apple.access_screensharing),12(everyone),33(_appstore),61(localaccounts),79(_appserverusr),80(admin),81(_appserveradm),98(_lpadmin),100(_lpoperator),204(_developer),402(com.apple.sharepoint.group.1)
^C
Abort session 4? [y/N]  y
```

Had I not noticed the carriage return / line feed come through, I would not be aware a session was returned.

Fortunately, attempting to CTRL+C the running handler reveals the presence of a session. Observe:

```
msf5 exploit(multi/handler) > run

[*] Started reverse TCP handler on 10.1.1.197:1337 

^C
Abort session 5? [y/N]  
```

I've performed no root cause analysis and don't intend to. At the time, I remember thinking 2c92e854942b24f3448acee1fe04a9b961a6c099 was the obvious culprit, but have not verified.

### PR DESCRIPTION
**Pull request by Green-m**
Fix #10129, no session report and `interact` crashed when database disabled.

This fix probably did not fix flawlessly, glad to receive any suggestion!

Maybe related: #9979 

### Before 

```
$ ./msfconsole -nq
[-] ***
[-] * WARNING: Database support has been disabled
[-] ***
[*] Starting persistent handler(s)...
msf5 auxiliary(scanner/ssh/libssh_auth_bypass) > options

Module options (auxiliary/scanner/ssh/libssh_auth_bypass):

   Name          Current Setting  Required  Description
   ----          ---------------  --------  -----------
   CHECK_BANNER  true             no        Check banner for libssh
   CMD                            no        Command to execute
   RHOSTS                         yes       The target address range or CIDR identifier
   RPORT         22               yes       The target port
   SPAWN_PTY     false            no        Spawn a PTY
   THREADS       1                yes       The number of concurrent threads

msf5 auxiliary(scanner/ssh/libssh_auth_bypass) > set rhosts 127.0.0.1
rhosts => 127.0.0.1
msf5 auxiliary(scanner/ssh/libssh_auth_bypass) > set rport 2222
rport => 2222
msf5 auxiliary(scanner/ssh/libssh_auth_bypass) > set spawn_pty true
spawn_pty => true
msf5 auxiliary(scanner/ssh/libssh_auth_bypass) > run

[*] 127.0.0.1:2222 - Attempting authentication bypass
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf5 auxiliary(scanner/ssh/libssh_auth_bypass) > sessions

Active sessions
===============

  Id  Name  Type   Information                                                  Connection
  --  ----  ----   -----------                                                  ----------
  1         shell   libssh Authentication Bypass Scanner (SSH-2.0-libssh_0.8.1)  127.0.0.1:62616 -> 127.0.0.1:2222 (127.0.0.1)

msf5 auxiliary(scanner/ssh/libssh_auth_bypass) > sessions -i 1
[*] Starting interaction with 1...

[-] Session manipulation failed: Problem reporting session event: No registered data_service. See log for more details. ["/Users/green/msfdev/metasploit-framework/lib/metasploit/framework/data_service/proxy/core.rb:158:in `log_error'", "/Users/green/msfdev/metasploit-framework/lib/metasploit/framework/data_service/proxy/session_event_data_proxy.rb:8:in `rescue in report_session_event'", "/Users/green/msfdev/metasploit-framework/lib/metasploit/framework/data_service/proxy/session_event_data_proxy.rb:3:in `report_session_event'", "/Users/green/msfdev/metasploit-framework/lib/msf/core/framework.rb:487:in `block in on_session_output'", "/Users/green/msfdev/metasploit-framework/lib/msf/core/framework.rb:485:in `each'", "/Users/green/msfdev/metasploit-framework/lib/msf/core/framework.rb:485:in `on_session_output'", "/Users/green/msfdev/metasploit-framework/lib/msf/core/event_dispatcher.rb:183:in `block in method_missing'", "/Users/green/msfdev/metasploit-framework/lib/msf/core/event_dispatcher.rb:181:in `each'", "/Users/green/msfdev/metasploit-framework/lib/msf/core/event_dispatcher.rb:181:in `method_missing'", "/Users/green/msfdev/metasploit-framework/lib/msf/base/sessions/command_shell.rb:492:in `shell_read'", "/Users/green/msfdev/metasploit-framework/lib/msf/base/sessions/command_shell.rb:611:in `_interact_stream'", "/Users/green/msfdev/metasploit-framework/lib/msf/base/sessions/command_shell.rb:598:in `_interact'", "/Users/green/msfdev/metasploit-framework/lib/rex/ui/interactive.rb:51:in `interact'", "/Users/green/msfdev/metasploit-framework/lib/msf/ui/console/command_dispatcher/core.rb:1362:in `cmd_sessions'", "/Users/green/msfdev/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:502:in `run_command'", "/Users/green/msfdev/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:453:in `block in run_single'", "/Users/green/msfdev/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:447:in `each'", "/Users/green/msfdev/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:447:in `run_single'", "/Users/green/msfdev/metasploit-framework/lib/rex/ui/text/shell.rb:151:in `run'", "/Users/green/msfdev/metasploit-framework/lib/metasploit/framework/command/console.rb:48:in `start'", "/Users/green/msfdev/metasploit-framework/lib/metasploit/framework/command/base.rb:82:in `start'", "./msfconsole:49:in `<main>'"]

```

### After the fix 

```
 ./msfconsole -nq
[-] ***
[-] * WARNING: Database support has been disabled
[-] ***
[*] Starting persistent handler(s)...
msf5 auxiliary(scanner/ssh/libssh_auth_bypass) > options

Module options (auxiliary/scanner/ssh/libssh_auth_bypass):

   Name          Current Setting  Required  Description
   ----          ---------------  --------  -----------
   CHECK_BANNER  true             no        Check banner for libssh
   CMD                            no        Command to execute
   RHOSTS                         yes       The target address range or CIDR identifier
   RPORT         22               yes       The target port
   SPAWN_PTY     false            no        Spawn a PTY
   THREADS       1                yes       The number of concurrent threads

msf5 auxiliary(scanner/ssh/libssh_auth_bypass) > set rhosts 127.0.0.1
rhosts => 127.0.0.1
msf5 auxiliary(scanner/ssh/libssh_auth_bypass) > set rport 2222
rport => 2222
msf5 auxiliary(scanner/ssh/libssh_auth_bypass) > set spawn_pty true
spawn_pty => true
msf5 auxiliary(scanner/ssh/libssh_auth_bypass) > run

[*] 127.0.0.1:2222 - Attempting authentication bypass
[*] Command shell session 1 opened (127.0.0.1:62598 -> 127.0.0.1:2222) at 2018-10-24 14:54:01 +0800
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf5 auxiliary(scanner/ssh/libssh_auth_bypass) > sessions -i 1
[*] Starting interaction with 1...

# id
id
uid=0(root) gid=0(root) groups=0(root)
# whoami
whoami
root
# ^Z
Background session 1? [y/N]  y
msf5 auxiliary(scanner/ssh/libssh_auth_bypass) >
```

### CODE REVIEW

File: lib/metasploit/framework/data_service/proxy/session_data_proxy.rb [hunk 1]
```diff
@@ -10,6 +10,10 @@ def sessions(opts={})
   end
 
   def report_session(opts)
+    unless self.active
```
Comment (wvu-r7) [inline lib/metasploit/framework/data_service/proxy/session_data_proxy.rb]: You can do `return unless self.active` for style consistency. :)

**Comment by mkienow-r7** [comment]
I think we need to dig into this a little further. This simply adds a check of `active` to the data proxy, but both of those call a DBManager method (`report_session` and `report_session_event`) that does the same:
```
return if not active
```

**Comment by Green-m** [comment]
@mkienow-r7 Thanks for your reply!

Actually, the bug occured in `report_session` and `report_session_event` when call the function `get_data_service`.

```
# metasploit-framework/lib/metasploit/framework/data_service/proxy/session_data_proxy.rb
  def report_session(opts)
    begin
      data_service = self.get_data_service
      add_opts_workspace(opts)
      data_service.report_session(opts)
    rescue => e
      self.log_error(e, "Problem reporting session")
    end
```

The `get_data_service` would raise an error directly if no database is connected, then it would be caught by the `rescue`, and `log_error` would raise an error too, which lead the whole function to fail.

I am not sure whether catching the exception in `get_data_service` to fix it is a good approach, the `raise` inside seems deliberate, so I fix it in `report_session` and `report_session_event`.

Maybe it's better to do it in `get_data_service` to catch all the same type exception, not for sure.

```
From: /Users/green/msfdev/metasploit-framework/lib/metasploit/framework/data_service/proxy/core.rb @ line 149 Metasploit::Framework::DataService::DataProxy#get_data_service:

    148: def get_data_service
 => 149:   raise 'No registered data_service' unless @current_data_service
    150:   return @current_data_service
    151: end
```

All stack trace:
```
[10/23/2018 18:35:42] [e(0)] core: Problem reporting session: No registered data_service
[10/23/2018 18:35:42] [e(0)] core: /Users/green/msfdev/metasploit-framework/lib/metasploit/framework/data_service/proxy/core.rb:149:in `get_data_service'
[10/23/2018 18:35:42] [e(0)] core: /Users/green/msfdev/metasploit-framework/lib/metasploit/framework/data_service/proxy/session_data_proxy.rb:14:in `report_session'
[10/23/2018 18:35:42] [e(0)] core: /Users/green/msfdev/metasploit-framework/lib/msf/core/framework.rb:419:in `on_session_open'
[10/23/2018 18:35:42] [e(0)] core: /Users/green/msfdev/metasploit-framework/lib/msf/core/event_dispatcher.rb:183:in `block in method_missing'
[10/23/2018 18:35:42] [e(0)] core: /Users/green/msfdev/metasploit-framework/lib/msf/core/event_dispatcher.rb:181:in `each'
[10/23/2018 18:35:42] [e(0)] core: /Users/green/msfdev/metasploit-framework/lib/msf/core/event_dispatcher.rb:181:in `method_missing'
[10/23/2018 18:35:42] [e(0)] core: /Users/green/msfdev/metasploit-framework/lib/msf/core/session_manager.rb:225:in `register'
[10/23/2018 18:35:42] [e(0)] core: /Users/green/msfdev/metasploit-framework/lib/msf/core/auxiliary/commandshell.rb:52:in `start_session'
[10/23/2018 18:35:42] [e(0)] core: /Users/green/msfdev/metasploit-framework/modules/auxiliary/scanner/ssh/libssh_auth_bypass.rb:127:in `run_host'
[10/23/2018 18:35:42] [e(0)] core: /Users/green/msfdev/metasploit-framework/lib/msf/core/auxiliary/scanner.rb:111:in `block (2 levels) in run'
[10/23/2018 18:35:42] [e(0)] core: /Users/green/msfdev/metasploit-framework/lib/msf/core/thread_manager.rb:100:in `block in spawn'
[10/23/2018 18:35:42] [w(0)] core: Exception in on_session_open event handler: Exception: Problem reporting session: No registered data_service. See log for more details.
[10/23/2018 18:35:42] [w(0)] core: Call Stack
/Users/green/msfdev/metasploit-framework/lib/metasploit/framework/data_service/proxy/core.rb:158:in `log_error'
/Users/green/msfdev/metasploit-framework/lib/metasploit/framework/data_service/proxy/session_data_proxy.rb:18:in `rescue in report_session'
/Users/green/msfdev/metasploit-framework/lib/metasploit/framework/data_service/proxy/session_data_proxy.rb:12:in `report_session'
/Users/green/msfdev/metasploit-framework/lib/msf/core/framework.rb:419:in `on_session_open'
/Users/green/msfdev/metasploit-framework/lib/msf/core/event_dispatcher.rb:183:in `block in method_missing'
/Users/green/msfdev/metasploit-framework/lib/msf/core/event_dispatcher.rb:181:in `each'
/Users/green/msfdev/metasploit-framework/lib/msf/core/event_dispatcher.rb:181:in `method_missing'
/Users/green/msfdev/metasploit-framework/lib/msf/core/session_manager.rb:225:in `register'
/Users/green/msfdev/metasploit-framework/lib/msf/core/auxiliary/commandshell.rb:52:in `start_session'
/Users/green/msfdev/metasploit-framework/modules/auxiliary/scanner/ssh/libssh_auth_bypass.rb:127:in `run_host'
/Users/green/msfdev/metasploit-framework/lib/msf/core/auxiliary/scanner.rb:111:in `block (2 levels) in run'
/Users/green/msfdev/metasploit-framework/lib/msf/core/thread_manager.rb:100:in `block in spawn'
```

**Comment by mkienow-r7** [comment]
@Green-m Thank you for pointing out the issue stemmed from the `get_data_service` method call. I'm looking into this closer and will followup shortly.

**Comment by mkienow-r7** [comment]
@Green-m I've opened a PR to your branch to enhance your fix in a manner that provides flexibility for the future in handling data service operations. I talked through this solution with @jbarnett-r7 another developer that has spent a good amount of time working in the new data services code. All of the `DataProxy` modules should be updated to use this pattern.

**Comment by bcoles** [comment]
Tested the original patch. Works for me.

```
root@kali:/pentest/exploit/metasploit-framework# ./msfconsole 
[-] ***rting the MetasPloit Framework console...-
[-] * WARNING: No database support: No database YAML file
[-] ***
                                                  

  PPPPP   IIIIIII   N    N
  P   PP     I      NN   N   IDENTIFICATION
  P   PP     I      N N  N
  PPPPP      I      N  N N      PROGRAM
  P          I      N   NN
  P       IIIIIII   N    N

  Strike a key when ready ...



       =[ metasploit v5.0.0-dev-eb5ab55                   ]
+ -- --=[ 1835 exploits - 1038 auxiliary - 317 post       ]
+ -- --=[ 539 payloads - 42 encoders - 10 nops            ]
+ -- --=[ 2 evasion                                       ]
+ -- --=[ ** This is Metasploit 5 development branch **   ]

msf5 > use exploit/multi/handler 
msf5 exploit(multi/handler) > set payload cmd/unix/reverse_netcat
payload => cmd/unix/reverse_netcat
msf5 exploit(multi/handler) > set lport 1337
lport => 1337
msf5 exploit(multi/handler) > set lhost 0.0.0.0
lhost => 0.0.0.0
msf5 exploit(multi/handler) > run

[*] Started reverse TCP handler on 0.0.0.0:1337 
[*] Command shell session 1 opened (127.0.0.1:1337 -> 127.0.0.1:34606) at 2018-10-25 22:01:14 -0400

uname 
Linux

```

**Comment by bcoles** [comment]
Mmmm.... nope. Those latest changes didn't work.

```
msf5 exploit(multi/handler) > run

[*] Started reverse TCP handler on 172.16.191.188:1337 

id
uid=0(root) gid=0(root) groups=0(root)
^C
Abort session 4? [y/N]  y
""

[*] 172.16.191.188 - Command shell session 4 closed.  Reason: User exit
```

**Comment by mkienow-r7** [comment]
@bcoles Thanks for the update. It appears there are two difference states that can be created when disconnected from the database, that is either `msfconsole -n` or `msfconsole` where when database.yml exists.

**Comment by mkienow-r7** [comment]
@Green-m I opened another small PR against your branch so that the enhancement resolves the issue when `msfconsole` is started without a `~/.msf4/database.yml` file. In the testing for my original PR I only tested by starting `msfconsole -n`.

**Comment by bcoles** [comment]
That appears to have fixed it.


```
[-] ***rting the Metasploit Framework console...|
[-] * WARNING: No database support: No database YAML file
[-] ***
[*] Starting the MetasPloit Framework console...-
                                                  
  LOGON: Help Logon

  HELP NOT AVAILABLE


  LOGON: Help Games

  'GAMES' REFERS TO MODELS, SIMULATIONS AND GAMES
  WHICH HAVE TACTICAL AND STRATEGIC APPLICATIONS.

  List of Games
  FALKEN'S MAZE
  BLACK JACK           .=======================.
  GIN RUMMY           |                         |
  HEARTS              |  SHALL WE PLAY A GAME ? |
  BRIDGE              |                         |
  CHECKERS             '======================='
  CHESS
  POKER
  FIGHTER COMBAT
  GUERRILLA ENGAGEMENT
  DESERT WARFARE
  AIR-TO-GROUND ACTIONS
  THEATERWIDE TACTICAL WARFARE
  THEATERWIDE BIOTOXIC AND CHEMICAL WARFARE

  GLOBAL THERMONUCLEAR WAR

  LOGON: JOSHUA


       =[ metasploit v5.0.0-dev-eb5ab55                   ]
+ -- --=[ 1835 exploits - 1038 auxiliary - 317 post       ]
+ -- --=[ 539 payloads - 42 encoders - 10 nops            ]
+ -- --=[ 2 evasion                                       ]
+ -- --=[ ** This is Metasploit 5 development branch **   ]

msf5 > 
msf5 > use exploit/multi/handler 
msf5 exploit(multi/handler) > set payload cmd/unix/reverse_netcat
payload => cmd/unix/reverse_netcat
msf5 exploit(multi/handler) > set lport 1337
lport => 1337
msf5 exploit(multi/handler) > set lhost 0.0.0.0
lhost => 0.0.0.0
msf5 exploit(multi/handler) > run

[*] Started reverse TCP handler on 0.0.0.0:1337 
[*] Command shell session 1 opened (127.0.0.1:1337 -> 127.0.0.1:34687) at 2018-10-26 00:08:00 -0400

id
uid=0(root) gid=0(root) groups=0(root)
^C
Abort session 1? [y/N]  y
""

[*] 127.0.0.1 - Command shell session 1 closed.  Reason: User exit
msf5 exploit(multi/handler) > 
```

**Comment by mkienow-r7** [comment]
@jbarnett-r7 Thank you for taking over and landing since I contributed code changes.

**Comment by jbarnett-r7** [comment]
# Release Notes

Fixes an issue that was preventing new session connections from displaying in the console when no database was connected.

**Pull request merged**